libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp

```diff
@@ -0,0 +1,201 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ *       AP_OpticalFlow_Linux.cpp - ardupilot library for the PX4Flow sensor.
+ *          inspired by the PX4Firmware code.
+ *      
+ *       @author: Víctor Mayoral Vilches
+ *
+ */
+
+#include <AP_HAL.h>
+#include "OpticalFlow.h"
+
+#define PX4FLOW_DEBUG 1
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#define PX4FLOW_I2C_ADDRESS     0x42    // 7-bit address. 8-bit address is 0x84, range 0x42 - 0x49
+#define PX4FLOW_REG             0x16    // Measure Register 22
+#define I2C_FRAME_SIZE          (sizeof(i2c_frame))
+#define I2C_INTEGRAL_FRAME_SIZE (sizeof(i2c_integral_frame))
+
+extern const AP_HAL::HAL& hal;
+
+AP_OpticalFlow_Linux::AP_OpticalFlow_Linux(OpticalFlow &_frontend) : 
+    OpticalFlow_backend(_frontend)
+{}
+
+void AP_OpticalFlow_Linux::init(void)
+{
+    // only initialise once
+    if (initialised) {
+        return;
+    }
+
+    // get pointer to i2c bus semaphore
+    AP_HAL::Semaphore *i2c_sem = hal.i2c->get_semaphore();
+    if (i2c_sem == NULL) {
+        return;
+    }
+
+    // take i2c bus sempahore
+    if (!i2c_sem->take(200)) {
+        return;
+    }
+
+    // read from flow sensor to ensure it is not a ll40ls Lidar (which can be on 0x42)
+    // read I2C_FRAME_SIZE bytes, the ll40ls will error whereas the flow happily returns data
+    uint8_t val[I2C_FRAME_SIZE];
+    if (hal.i2c->readRegisters(PX4FLOW_I2C_ADDRESS, 0, I2C_FRAME_SIZE, val) != 0) {
+        i2c_sem->give();
+        return;
+    }
+
+    // success
+    initialised = true;
+    i2c_sem->give();
+}
+
+bool AP_OpticalFlow_Linux::request_measurement()
+{
+    // send measure request to sensor
+    uint8_t cmd = PX4FLOW_REG;
+    if (hal.i2c->writeRegisters(PX4FLOW_I2C_ADDRESS, cmd, 0, nullptr) != 0) {
+        return false;
+    }
+    return true;
+}
+
+bool AP_OpticalFlow_Linux::read(optical_flow_s* report)
+{
+    // get pointer to i2c bus semaphore
+    AP_HAL::Semaphore *i2c_sem = hal.i2c->get_semaphore();
+    if (i2c_sem == NULL) {
+        num_errors++;
+        return false;
+    }
+
+    // take i2c bus sempahore (non blocking)
+    if (!i2c_sem->take_nonblocking()) {
+        num_errors++;
```

Comment:
hummn... I'm confused about these `num_errors`  handling.  Suppose that for an instance the i2c bus got way too busy with other things and thus take_nonblocking() returns false here.  You'd never try again to read this sensor again because `update()` does:
```
    if (!initialised || num_errors >= 20) {
        return;
    }
```

and in that case num_errors would never be decremented.  I think we shouldn't increment num_errors here.